File: backend/src/main/java/com/agentplatform/backend/runs/RunService.java
package com.agentplatform.backend.runs;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import java.time.Instant;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
public class RunService {

    private final RunRepository runRepository;
    private final RunEventRepository runEventRepository;

    public RunService(RunRepository runRepository, RunEventRepository runEventRepository) {
        this.runRepository = runRepository;
        this.runEventRepository = runEventRepository;
    }

    public RunResponse createRun(Long ownerId, RunRequest request) {
        if (request.getTaskId() == null) {
            throw new IllegalArgumentException("Task ID is required");
        }
        if (request.getAgentId() == null) {
            throw new IllegalArgumentException("Agent ID is required");
        }

        Run run = new Run(request.getTaskId(), request.getAgentId(), ownerId);
        run.setInput(request.getInput());
        if (request.getStatus() != null) {
            run.setStatus(request.getStatus());
        }

        Run saved = runRepository.save(run);
        return new RunResponse(saved);
    }

    public Optional<RunResponse> getRunById(Long id, Long ownerId) {
        return runRepository.findByIdAndOwnerId(id, ownerId)
                .map(RunResponse::new);
    }

    public Page<RunResponse> listRuns(Long ownerId, Pageable pageable) {
        return runRepository.findByOwnerId(ownerId, pageable)
                .map(RunResponse::new);
    }

    public Page<RunResponse> listRunsByStatus(Long ownerId, String status, Pageable pageable) {
        return runRepository.findByOwnerIdAndStatus(ownerId, status, pageable)
                .map(RunResponse::new);
    }

    public Page<RunResponse> listRunsByTask(Long ownerId, Long taskId, Pageable pageable) {
        return runRepository.findByOwnerIdAndTaskId(ownerId, taskId, pageable)
                .map(RunResponse::new);
    }

    public Page<RunResponse> listRunsByAgent(Long ownerId, Long agentId, Pageable pageable) {
        return runRepository.findByOwnerIdAndAgentId(ownerId, agentId, pageable)
                .map(RunResponse::new);
    }

    public RunResponse updateRunStatus(Long id, Long ownerId, String newStatus) {
        Run run = runRepository.findByIdAndOwnerId(id, ownerId)
                .orElseThrow(() -> new IllegalArgumentException("Run not found or not owned by user"));

        run.setStatus(newStatus);
        run.setUpdatedAt(Instant.now());

        Run updated = runRepository.save(run);
        return new RunResponse(updated);
    }

    public RunResponse setRunStarted(Long id, Long ownerId) {
        Run run = runRepository.findByIdAndOwnerId(id, ownerId)
                .orElseThrow(() -> new IllegalArgumentException("Run not found or not owned by user"));

        run.setStatus("running");
        run.setStartedAt(Instant.now());
        run.setUpdatedAt(Instant.now());

        Run updated = runRepository.save(run);
        return new RunResponse(updated);
    }

    public RunResponse completeRun(Long id, Long ownerId, String output, String logs) {
        Run run = runRepository.findByIdAndOwnerId(id, ownerId)
                .orElseThrow(() -> new IllegalArgumentException("Run not found or not owned by user"));

        run.setStatus("completed");
        run.setOutput(output);
        run.setLogs(logs);
        run.setCompletedAt(Instant.now());
        run.setUpdatedAt(Instant.now());

        Run updated = runRepository.save(run);
        return new RunResponse(updated);
    }

    public RunResponse failRun(Long id, Long ownerId, String errorMessage, String logs) {
        Run run = runRepository.findByIdAndOwnerId(id, ownerId)
                .orElseThrow(() -> new IllegalArgumentException("Run not found or not owned by user"));

        run.setStatus("failed");
        run.setErrorMessage(errorMessage);
        run.setLogs(logs);
        run.setCompletedAt(Instant.now());
        run.setUpdatedAt(Instant.now());

        Run updated = runRepository.save(run);
        return new RunResponse(updated);
    }

    public void cancelRun(Long id, Long ownerId) {
        Run run = runRepository.findByIdAndOwnerId(id, ownerId)
                .orElseThrow(() -> new IllegalArgumentException("Run not found or not owned by user"));

        if ("running".equals(run.getStatus())) {
            run.setStatus("cancelled");
            run.setCompletedAt(Instant.now());
            run.setUpdatedAt(Instant.now());
            runRepository.save(run);
        }
    }

    public long countRunsByStatus(Long ownerId, String status) {
        return runRepository.countByOwnerIdAndStatus(ownerId, status);
    }

    public List<RunResponse> getRecentRunsByStatus(Long ownerId, String status, int limit) {
        return runRepository.findByOwnerIdAndStatusOrderByCreatedAtDesc(ownerId, status).stream()
                .limit(limit)
                .map(RunResponse::new)
                .collect(Collectors.toList());
    }

    public RunResponse toResponse(Run run) {
        return new RunResponse(run);
    }

    // ==================== Event Handling ====================

    public RunEvent recordEvent(Long runId, String eventType, String payload) {
        long sequence = runEventRepository.countByRunId(runId);
        RunEvent event = new RunEvent(runId, sequence + 1, eventType, payload);
        return runEventRepository.save(event);
    }

    public List<RunEvent> getRunEvents(Long runId) {
        return runEventRepository.findByRunIdOrderBySequenceAsc(runId);
    }

    public long getEventCount(Long runId) {
        return runEventRepository.countByRunId(runId);
    }
}

File: frontend/agentplatform/src/app/services/run.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface Run {
  id: number;
  taskId: number;
  agentId: number;
  ownerId: number;
  status: string;
  input?: string;
  output?: string;
  logs?: string;
  errorMessage?: string;
  startedAt?: string;
  completedAt?: string;
  createdAt: string;
  updatedAt: string;
}

export interface RunRequest {
  taskId: number;
  agentId: number;
  input?: string;
  status?: string;
}

@Injectable({
  providedIn: 'root'
})
export class RunService {
  private apiUrl = '/api/runs';

  constructor(private http: HttpClient) { }

  /**
   * Start a new run for an agent
   */
  startAgentRun(agentId: number, taskId?: number, description?: string): Observable<Run> {
    const payload = {
      agentId,
      taskId: taskId || null,
      description: description || 'Run started from dashboard'
    };
    return this.http.post<Run>(`/api/agents/${agentId}/runs`, payload);
  }

  /**
   * List runs for a specific agent
   */
  listAgentRuns(agentId: number): Observable<Run[]> {
    return this.http.get<Run[]>(`/api/agents/${agentId}/runs`);
  }

  createRun(request: RunRequest): Observable<Run> {
    return this.http.post<Run>(this.apiUrl, request);
  }

  getRun(id: number): Observable<Run> {
    return this.http.get<Run>(`${this.apiUrl}/${id}`);
  }

  listRuns(page: number = 0, size: number = 20): Observable<any> {
    let params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
    return this.http.get<any>(this.apiUrl, { params });
  }

  listRunsByStatus(status: string, page: number = 0, size: number = 20): Observable<any> {
    let params = new HttpParams()
      .set('status', status)
      .set('page', page.toString())
      .set('size', size.toString());
    return this.http.get<any>(`${this.apiUrl}/filter/status`, { params });
  }

  listRunsByTask(taskId: number, page: number = 0, size: number = 20): Observable<any> {
    let params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
    return this.http.get<any>(`${this.apiUrl}/filter/task/${taskId}`, { params });
  }

  listRunsByAgent(agentId: number, page: number = 0, size: number = 20): Observable<any> {
    let params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
    return this.http.get<any>(`${this.apiUrl}/filter/agent/${agentId}`, { params });
  }

  updateRunStatus(id: number, status: string): Observable<Run> {
    let params = new HttpParams().set('status', status);
    return this.http.put<Run>(`${this.apiUrl}/${id}/status`, {}, { params });
  }

  startRun(id: number): Observable<Run> {
    return this.http.post<Run>(`${this.apiUrl}/${id}/start`, {});
  }

  completeRun(id: number, output?: string, logs?: string): Observable<Run> {
    let params = new HttpParams();
    if (output) params = params.set('output', output);
    if (logs) params = params.set('logs', logs);
    return this.http.post<Run>(`${this.apiUrl}/${id}/complete`, {}, { params });
  }

  failRun(id: number, errorMessage?: string, logs?: string): Observable<Run> {
    let params = new HttpParams();
    if (errorMessage) params = params.set('errorMessage', errorMessage);
    if (logs) params = params.set('logs', logs);
    return this.http.post<Run>(`${this.apiUrl}/${id}/fail`, {}, { params });
  }

  cancelRun(id: number): Observable<void> {
    return this.http.post<void>(`${this.apiUrl}/${id}/cancel`, {});
  }

  countRunsByStatus(status: string): Observable<number> {
    let params = new HttpParams().set('status', status);
    return this.http.get<number>(`${this.apiUrl}/stats/count`, { params });
  }
}
